drivers/iio/adc/ad4630.c

```
@@ -0,0 +1,1185 @@
+
+		ret = ad4630_get_chan_gain(st, chan->scan_index, &temp);
+		iio_device_release_direct_mode(indio_dev);
+		if (ret)
+			return ret;
+
+		*val = temp / 10000;
+		*val2 = (temp - (*val * 10000)) * 100;
+
+		return IIO_VAL_INT_PLUS_MICRO;
+	case IIO_CHAN_INFO_OFFSET:
+		ret = iio_device_claim_direct_mode(indio_dev);
+		if (ret)
+			return ret;
+
+		ret = ad4630_get_chan_offset(st, chan->scan_index, val);
+		iio_device_release_direct_mode(indio_dev);
+		if (ret)
+			return ret;
+
+		return IIO_VAL_INT;
+	default:
+		return -EINVAL;
+	}
+}
+
+static int ad4630_write_raw(struct iio_dev *indio_dev,
+			    struct iio_chan_spec const *chan, int val,
+			    int val2, long info)
+{
+	struct ad4630_state *st = iio_priv(indio_dev);
+	int ret;
+
+	switch (info) {
+	case IIO_CHAN_INFO_SAMP_FREQ:
+		return ad4630_set_sampling_freq(st, val);
+	case IIO_CHAN_INFO_HARDWAREGAIN:
+		ret = iio_device_claim_direct_mode(indio_dev);
+		if (ret)
+			return ret;
+		ret = ad4630_set_chan_gain(st, chan->scan_index, val, val2);
+		iio_device_release_direct_mode(indio_dev);
+
+		return ret;
+	case IIO_CHAN_INFO_OFFSET:
+		ret = iio_device_claim_direct_mode(indio_dev);
+		if (ret)
+			return ret;
+		ret = ad4630_set_chan_offset(st, chan->scan_index, val);
+		iio_device_release_direct_mode(indio_dev);
+
+		return ret;
+	default:
+		return -EINVAL;
+	}
+}
+
+static int ad4630_dma_buffer_submit_block(struct iio_dma_buffer_queue *queue,
+					  struct iio_dma_buffer_block *block)
+{
+	return iio_dmaengine_buffer_submit_block(queue, block, DMA_DEV_TO_MEM);
+}
+
+static int ad4630_buffer_preenable(struct iio_dev *indio_dev)
+{
+	struct ad4630_state *st = iio_priv(indio_dev);
+	unsigned int spi_rx_data[4];
+	struct spi_transfer xfer = {
+		/* If tx_buf is empty, the spi engine won't write on SDO_FIFO */
+		.tx_buf = NULL,
+		.rx_buf = spi_rx_data,
+		.len = 1,
+		.speed_hz = AD4630_SPI_SAMPLING_SPEED,
+	};
+	struct spi_message msg;
+	int width;
+	int ret;
+
+	pm_runtime_get_sync(&st->spi->dev);
```

This function might fail in which case I guess we should also return an error. Loot at [pm_runtime_resume_and_get()](https://elixir.bootlin.com/linux/v5.10/C/ident/pm_runtime_resume_and_get) to see how to handle errors. Or just use it directly :)